File: bindings/lua/Polycode/Ray.lua
class "Ray"


function Ray:__getvar(name)
	if name == "origin" then
		local retVal = Polycode.Ray_get_origin(self.__ptr)
		if retVal == nil then return nil end
		local __c = _G["Vector3"]("__skip_ptr__")
		__c.__ptr = retVal
		return __c
	elseif name == "direction" then
		local retVal = Polycode.Ray_get_direction(self.__ptr)
		if retVal == nil then return nil end
		local __c = _G["Vector3"]("__skip_ptr__")
		__c.__ptr = retVal
		return __c
	elseif name == "inv_direction" then
		local retVal = Polycode.Ray_get_inv_direction(self.__ptr)
		if retVal == nil then return nil end
		local __c = _G["Vector3"]("__skip_ptr__")
		__c.__ptr = retVal
		return __c
	end
end

function Ray:__setvar(name,value)
	if name == "origin" then
		Polycode.Ray_set_origin(self.__ptr, value.__ptr)
		return true
	elseif name == "direction" then
		Polycode.Ray_set_direction(self.__ptr, value.__ptr)
		return true
	elseif name == "inv_direction" then
		Polycode.Ray_set_inv_direction(self.__ptr, value.__ptr)
		return true
	end
	return false
end
function Ray:Ray(...)
	local arg = {...}
	for k,v in pairs(arg) do
		if type(v) == "table" then
			if v.__ptr ~= nil then
				arg[k] = v.__ptr
			end
		end
	end
	if self.__ptr == nil and arg[1] ~= "__skip_ptr__" then
		self.__ptr = Polycode.Ray(unpack(arg))
	end
end

function Ray:boxIntersect(box, transformMatrix, vnear, vfar)
	local retVal = Polycode.Ray_boxIntersect(self.__ptr, box.__ptr, transformMatrix.__ptr, vnear, vfar)
	return retVal
end

function Ray:planeIntersectPoint(planeNormal, planeDistance)
	local retVal = Polycode.Ray_planeIntersectPoint(self.__ptr, planeNormal.__ptr, planeDistance)
	if retVal == nil then return nil end
	local __c = _G["Vector3"]("__skip_ptr__")
	__c.__ptr = retVal
	return __c
end

function Ray:tranformByMatrix(matrix)
	local retVal = Polycode.Ray_tranformByMatrix(self.__ptr, matrix.__ptr)
	if retVal == nil then return nil end
	local __c = _G["Ray"]("__skip_ptr__")
	__c.__ptr = retVal
	return __c
end

function Ray:closestPointOnRay(point)
	local retVal = Polycode.Ray_closestPointOnRay(self.__ptr, point.__ptr)
	if retVal == nil then return nil end
	local __c = _G["Vector3"]("__skip_ptr__")
	__c.__ptr = retVal
	return __c
end

function Ray:polygonIntersect(v1, v2, v3)
	local retVal = Polycode.Ray_polygonIntersect(self.__ptr, v1.__ptr, v2.__ptr, v3.__ptr)
	return retVal
end

function Ray:__delete()
	if self then Polycode.delete_Ray(self.__ptr) end
end

File: bindings/javascript/Polycode/Ray.js
function Ray() {
	if(arguments[0] != "__skip_ptr__") {
		this.__ptr = Polycode.Ray()
	}
	Object.defineProperties(this, {
		'origin': { enumerable: true, configurable: true, get: Ray.prototype.__get_origin, set: Ray.prototype.__set_origin},
		'direction': { enumerable: true, configurable: true, get: Ray.prototype.__get_direction, set: Ray.prototype.__set_direction},
		'inv_direction': { enumerable: true, configurable: true, get: Ray.prototype.__get_inv_direction, set: Ray.prototype.__set_inv_direction}
	})
}


Ray.prototype.__get_origin = function() {
	var retVal = new Vector3("__skip_ptr__")
	retVal.__ptr = 	Polycode.Ray__get_origin(this.__ptr)
	return retVal
}

Ray.prototype.__set_origin = function(val) {
	Polycode.Ray__set_origin(this.__ptr, val.__ptr)
}

Ray.prototype.__get_direction = function() {
	var retVal = new Vector3("__skip_ptr__")
	retVal.__ptr = 	Polycode.Ray__get_direction(this.__ptr)
	return retVal
}

Ray.prototype.__set_direction = function(val) {
	Polycode.Ray__set_direction(this.__ptr, val.__ptr)
}

Ray.prototype.__get_inv_direction = function() {
	var retVal = new Vector3("__skip_ptr__")
	retVal.__ptr = 	Polycode.Ray__get_inv_direction(this.__ptr)
	return retVal
}

Ray.prototype.__set_inv_direction = function(val) {
	Polycode.Ray__set_inv_direction(this.__ptr, val.__ptr)
}

Duktape.fin(Ray.prototype, function (x) {
	if (x === Ray.prototype) {
		return;
	}
	Polycode.Ray__delete(x.__ptr)
})

Ray.prototype.boxIntersect = function(box,transformMatrix,vnear,vfar) {
	return Polycode.Ray_boxIntersect(this.__ptr, box, transformMatrix, vnear, vfar)
}

Ray.prototype.planeIntersectPoint = function(planeNormal,planeDistance) {
	var retVal = new Vector3("__skip_ptr__")
	retVal.__ptr = Polycode.Ray_planeIntersectPoint(this.__ptr, planeNormal, planeDistance)
	return retVal
}

Ray.prototype.tranformByMatrix = function(matrix) {
	var retVal = new Ray("__skip_ptr__")
	retVal.__ptr = Polycode.Ray_tranformByMatrix(this.__ptr, matrix)
	return retVal
}

Ray.prototype.closestPointOnRay = function(point) {
	var retVal = new Vector3("__skip_ptr__")
	retVal.__ptr = Polycode.Ray_closestPointOnRay(this.__ptr, point)
	return retVal
}

Ray.prototype.polygonIntersect = function(v1,v2,v3) {
	return Polycode.Ray_polygonIntersect(this.__ptr, v1, v2, v3)
}
